case-studies/n-body/unity/N-body/Assets/Scripts/ECS/Systems/BenchmarkSystem.cs

```diff
@@ -1,37 +1,141 @@
+
+        if (!IsEcs)
         {
-            stopwatch.Start();
+            for (var n = 8; n <= LARGEST_EXPONENT; n++) benchmarks[n].TicksPerSample /= 10;
+            for (var n = 12; n <= LARGEST_EXPONENT; n++) benchmarks[n].TicksPerSample /= 10;
         }
-        else if (ticks == SIMULATED_TICKS)
+
+        CurrentBenchmark = benchmarks[0];
+    }
+
+    protected override void OnUpdate()
+    {
+        if (currentTick == 0) stopwatch.Start();
+        if (currentTick == CurrentBenchmark.TicksPerSample)
         {
             stopwatch.Stop();
-            var duration = stopwatch.Elapsed;
-            double tps = SIMULATED_TICKS / duration.TotalSeconds;
 
-            Debug.Log($"Benchmark took {math.round(duration.TotalMilliseconds)} ms to complete");
-            Debug.Log($"{(float) tps} avg tps over {SIMULATED_TICKS} ticks");
+            CurrentBenchmark.Samples[currentSample] = stopwatch.Elapsed;
+
+            if (++currentSample == SAMPLE_COUNT)
+            {
+                Debug.Log(
+                    $"Finished benchmark with {CurrentBenchmark.BodyCount} {(CurrentBenchmark.BodyCount > 1 ? "bodies" : "body")}.");
+
+                if (++currentBenchmarkIndex >= benchmarks.Length)
+                {
+                    Debug.Log("All benchmarks are done!");
+
+                    SaveAsCsv();
+
+                    if (SystemInfo.graphicsDeviceType == GraphicsDeviceType.Null) Application.Quit();
+                }
+                else
+                {
+                    CurrentBenchmark = benchmarks[currentBenchmarkIndex];
 
-            if (SystemInfo.graphicsDeviceType == GraphicsDeviceType.Null)
+                    ResetSimulation();
+
+                    currentSample = 0;
+                    currentTick = -2;
+                    stopwatch.Reset();
+                }
+            }
+            else
             {
-                Application.Quit();
+                currentTick = 0;
+                stopwatch.Restart();
             }
         }
 
-        ticks++;
+        currentTick++;
+    }
+
+    private void ResetSimulation()
+    {
+        if (IsEcs) ResetEcsSimulation();
+        else ResetGameObjectsSimulation();
+    }
+
+    private void ResetEcsSimulation()
+    {
+        var ecb = new EntityCommandBuffer(Allocator.TempJob);
+        Entities.ForEach((Entity entity, in Acceleration _) => ecb.DestroyEntity(entity)).Schedule();
+        Dependency.Complete();
+        ecb.Playback(EntityManager);
+        ecb.Dispose();
+
+        World.GetExistingSystemManaged<BodySpawnerSystem>().SpawnedBodies = false;
+    }
+
+    private void ResetGameObjectsSimulation()
+    {
+        foreach (var body in Object.FindObjectsOfType<Body>()) Object.Destroy(body.gameObject);
+
+        Object.FindObjectOfType<GameObjectBodySpawner>().SpawnBodies();
+    }
+
+    private void SaveAsCsv()
+    {
+        var path = Path.GetFullPath($"unity_{(IsEcs ? "ecs" : "gameobjects")}.csv");
```

Comment:
The path to `recs/benchmarking` cannot be hard coded relative to the executable, because the Unity executable can be placed anywhere and the EnTT executable will depend on from where CMake is called. Right now, it is writing to the working directory. For EnTT, `$ContentRoot$/benchmarking` could be used as the working directory from a CLion configuration. I could also add an environment variable for the output directory. What do you think?

Reply:
Alright then it's probably not worth the extra steps necessary to work with the environment variables. Then I'd say you keep placing them where they're being placed right now.

I think a better solution is to incorporate these tests (as much as possible, at least) into the benchmarking script I've created in #115. I'm thinking that we can have that script run each of the different benchmarks (Rust, EnTT, Unity) and then copy all of the `.csv`s into the `recs/benchmarking` directory, like I'm doing currently with the Rust-benchmarks. 

Do you think it's possible for you to call the benchmarks from there, and then copy the outputs like I'm doing?

Reply:
I think building the benchmarks would be a little tricky from a script. It is possible for both Unity and CMake to build from cli, but it would be harder to configure different toolchains and scripting backends.

Reply:
Alright, so we can at least automate the Unity-default scripting backend and EnTT compiled with CMake then. That's better than nothing!

I don't see how compiling with different toolchains is not possible through the CLI though. Once you've got them set up on your system, isn't it just a matter of invoking the different compilers through their cli?

Reply:
How I have it set up on Windows is that the toolchains are configured in CLion, and CLion is using a bundled CMake installation for generating the build files. The build system used to build depends on the toolchain and is usually a bundled version of Ninja from either CLion or Visual Studio.


Maybe it would be enough to allow the script to run all benchmarks in a folder of precompiled binaries instead.

Reply:
Ah, I see why that complicates things then.

> Maybe it would be enough to allow the script to run all benchmarks in a folder of precompiled binaries instead.

That sounds like a great idea!